libultraship/libultraship/HIDController.cpp

```diff
@@ -0,0 +1,369 @@
+    }
+
+    void HIDController::NormalizeStickAxis(int16_t wAxisValueX, int16_t wAxisValueY, int16_t wAxisThreshold) {
+        //scale {-32768 ... +32767} to {-84 ... +84}
+        auto ax = wAxisValueX * 85.0 / 32767.0;
+        auto ay = wAxisValueY * 85.0 / 32767.0;
+
+        //create scaled circular dead-zone in range {-15 ... +15}
+        auto len = sqrt(ax * ax + ay * ay);
+        if (len < wAxisThreshold) {
+            len = 0;
+        }
+        else if (len > 85.0) {
+            len = 85.0 / len;
+        }
+        else {
+            len = (len - wAxisThreshold) * 85.0 / (85.0 - wAxisThreshold) / len;
+        }
+        ax *= len;
+        ay *= len;
+
+        //bound diagonals to an octagonal range {-68 ... +68}
+        if (ax != 0.0 && ay != 0.0) {
+            auto slope = ay / ax;
+            auto edgex = copysign(85.0 / (abs(slope) + wAxisThreshold / 69.0), ax);
+            auto edgey = copysign(std::min(abs(edgex * slope), 85.0 / (1.0 / abs(slope) + wAxisThreshold / 69.0)), ay);
+            edgex = edgey / slope;
+
+            auto scale = sqrt(edgex * edgex + edgey * edgey) / 85.0;
+            ax *= scale;
+            ay *= scale;
+        }
+
+        wStickX = +ax;
+        wStickY = -ay;
+    }
+
+    void HIDController::ReadFromSource() {
+        std::string ConfSection = GetBindingConfSection();
+        std::shared_ptr<ConfigFile> pConf = GlobalCtx2::GetInstance()->GetConfig();
+        ConfigFile& Conf = *pConf.get();
+
+        SDL_GameControllerUpdate();
+
+        // If the controller is disconnected, close it.
+        if (Cont != nullptr && !SDL_JoystickGetAttached(Cont)) {
+            Close();
+        }
+
+        // Attempt to load the controller if it's not loaded
+        if (Cont == nullptr) {
+            // If we failed to load the controller, don't process it.
+            if (!Open()) {
+                return;
+            }
+        }
+
+        for (int32_t i = 0; i < numButtons; i++) {
+            if (ButtonMapping.contains(i)) {
+                if (SDL_JoystickGetButton(Cont, i)) {
+                    dwPressedButtons |= ButtonMapping[i];
+                }
+                else {
+                    dwPressedButtons &= ~ButtonMapping[i];
+                }
+            }
+        }
+
+        int StickAxisX = HID_CONTROLLER_AXIS_INVALID;
+        int StickAxisY = HID_CONTROLLER_AXIS_INVALID;
+        int32_t StickDeadzone = 0;
+
+        for (int32_t i = HID_CONTROLLER_AXIS_LEFTX; i < HID_CONTROLLER_AXIS_MAX; i++) {
+            auto Axis = i;
+            auto PosScancode = i + AXIS_SCANCODE_BIT;
+            auto NegScancode = -PosScancode;
+            auto AxisThreshold = ThresholdMapping[i];
+            auto PosButton = ButtonMapping[PosScancode];
+            auto NegButton = ButtonMapping[NegScancode];
+            auto AxisValue = SDL_JoystickGetAxis(Cont, Axis);
+
+#ifdef TARGET_WEB
```

Comment:
SoH doesn't have web as a target.

I didn't get a straight answer whether or not this should be included... So ¯\\\_(ツ)\_/¯

Reply:
I can remove this, the only thing I can think of is that this is for libultraship and not SoH. So depending on their future plans, this may want to be kept?

Reply:
I searched for target_web and didn't see it used anywhere else. Perhaps remove it. If it's ever needed it can be added back in.

Reply:
Sounds good. Done in [32a684b](https://github.com/HarbourMasters/Shipwright/pull/86/commits/32a684b9cd19e477220d95adf4ec1ff458ce5044)